### PR TITLE
Provide database based authorities

### PR DESCRIPTION
**Pull request by jcoyne**

### CODE REVIEW

File: lib/generators/qa/local/tables/tables_generator.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    include ActiveRecord::Generators::Migration
+
+    def migrations
+      generate "model qa/local_authority name:string:uniq"
+      generate "model qa/local_authority_entry local_authority:references label:string uri:string:uniq"
+      message = "Rails doesn't support functional indexes in migrations, so you'll have to add this manually:\n" \
+    "CREATE INDEX \"index_qa_local_authority_entries_on_lower_label\" ON \"qa_local_authority_entries\" (local_authority_id, lower(label))\n" \
+    "   OR on Sqlite: \n" \
+    "CREATE INDEX \"index_qa_local_authority_entries_on_lower_label\" ON \"qa_local_authority_entries\" (local_authority_id, label collate nocase)\n"
+      say_status("info", message, :yellow)
```
Comment (mjgiarlo) [inline lib/generators/qa/local/tables/tables_generator.rb]: Not `:blue`? Should we tag the UI/UX group?

;)

File: README.md [hunk 2]
```diff
@@ -170,7 +170,7 @@ For simple use cases when you have a few terms that don't change very often.
 
 Run the generator to install configuration files and an example authority.
 
-    rails generate qa:local
+    rails generate qa:local:files
```
Comment (mjgiarlo) [inline README.md]: Should we add a new section to the README for table-based authorities? That way we can reinforce what the generator spits out about having to manually create an index.

File: lib/qa/authorities/local/table_based_authority.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+          "   OR on Sqlite: \n" \
+          "CREATE INDEX \"index_qa_local_authority_entries_on_lower_label\" ON \"qa_local_authority_entries\" (local_authority_id, label collate nocase)\n" \
+      end
+    end
+
+    attr_reader :subauthority
+    def initialize(subauthority)
+      @subauthority = subauthority
+    end
+    def search(q)
```
Comment (mjgiarlo) [inline lib/qa/authorities/local/table_based_authority.rb]: Insert newline between methods?

File: lib/qa/authorities/local/table_based_authority.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      conn = ActiveRecord::Base.connection
+      if conn.table_exists?('local_authority_entries') && !conn.indexes('local_authority_entries').find { |i| i.name == 'index_local_authority_entries_on_lower_label' }
+        Rails.logger.error "You've installed local authority tables, but you haven't indexed the label.  Rails doesn't support functional indexes in migrations, so you'll have to add this manually:\n" \
+          "CREATE INDEX \"index_qa_local_authority_entries_on_lower_label\" ON \"qa_local_authority_entries\" (local_authority_id, lower(label))\n" \
+          "   OR on Sqlite: \n" \
+          "CREATE INDEX \"index_qa_local_authority_entries_on_lower_label\" ON \"qa_local_authority_entries\" (local_authority_id, label collate nocase)\n" \
+      end
+    end
+
+    attr_reader :subauthority
```
Comment (mjgiarlo) [inline lib/qa/authorities/local/table_based_authority.rb]: I'm thrown a bit by `subauthority`. Why not just `authority`?

File: spec/test_app_templates/lib/generators/test_app_generator.rb [hunk 5]
```diff
@@ -14,7 +14,8 @@ def run_qa_installer
   end
 
   def run_local_authority_installer
-    generate "qa:local"
+    generate "qa:local:files"
+    generate "qa:local:tables"
```
Comment (mjgiarlo) [inline spec/test_app_templates/lib/generators/test_app_generator.rb]: Does the test app not require the index to be added?

File: README.md (hunk 2 above)
Comment (jcoyne) [inline README.md]: yes, but I want to see if this all works with sufia first.

File: lib/qa/authorities/local/table_based_authority.rb (hunk 4 above)
Comment (jcoyne) [inline lib/qa/authorities/local/table_based_authority.rb]: This is following the pattern of the whole gem.  `Local` is the authority, and `<your-string-here>` is the sub-authority.

File: spec/test_app_templates/lib/generators/test_app_generator.rb (hunk 5 above)
Comment (jcoyne) [inline spec/test_app_templates/lib/generators/test_app_generator.rb]: The index is purely for performance.  You wouldn't notice if it does a whole table scan on a few rows.

File: README.md (hunk 2 above)
Comment (jcoyne) [inline README.md]: I've added a section to the readme on using the tables.

File: lib/qa/authorities/local/table_based_authority.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    def initialize(subauthority)
+      @subauthority = subauthority
+    end
+
+    def search(q)
+      return [] if q.blank?
+      output_set(base_relation.where('lower(label) like ?', "#{q.downcase}%").limit(25))
```
Comment (cbeer) [inline lib/qa/authorities/local/table_based_authority.rb]: Is `lower()` a SQL standard?

Comment (jcoyne) [inline lib/qa/authorities/local/table_based_authority.rb]: Every database that I know of supports it. However with SQLite you can't use it to make an index in this way.

Comment (jcoyne) [inline lib/qa/authorities/local/table_based_authority.rb]: Yep. ANSI SQL 2003

**Pull request merged**